File: server/tools/check_my_health.js
// The conversational read path into self-diagnosis (root CLAUDE.md's
// Operational Awareness item 1) — "has anything gone wrong with you
// lately?", "are you working correctly?" Runs every registered check live,
// right now, AND surfaces recent real history from the activity trace —
// distinct from check_environment.js (the surrounding MACHINE) and
// check_myself.js (Self-Model's own reliability/provenance/goal dimensions,
// nothing about malfunction or possible compromise).
//
// core:true (no reliable search-intent text) + meta:true
// (live-conversation-only). No confirm gate — every check here is
// read-only; the one thing that ever WRITES anything (a check's own
// remedy()) already ran, if at all, on the Heartbeat's own background tick,
// never triggered by this tool.

import { listChecks } from '../ops/diagnostics/registry.js';
import { listRecentForSource } from '../ops/ops-trace.js';

export default {
  name: 'check_my_health',
  core: true,
  meta: true,
  description:
    'Check whether you are actually working correctly right now, and whether anything real has gone wrong recently — Memory writing, ' +
    'background Jobs and the Scheduler actually progressing, the database itself, the voice pipeline\'s server-side half, and signs of ' +
    'possible tampering (an unexplained change to a secret/config file, a spike in auth failures or confirm-gate bypass attempts, an ' +
    'unexpected new local listener). This is about YOUR OWN malfunction or compromise, not the surrounding machine — use ' +
    'check_environment for that. Never answer "am I working fine" from impression — check here first.',
  parameters: {
    type: 'object',
    properties: {},
    required: [],
  },
  async run() {
    const checks = listChecks();
    const results = [];
    for (const c of checks) {
      let result;
      try {
        result = await c.probe();
      } catch (err) {
        result = { ok: false, detail: `Probe threw: ${err?.message || err}` };
      }
      results.push({ id: c.id, ok: result.ok !== false, detail: result.detail || null });
    }

    const failing = results.filter((r) => !r.ok);
    const recentHistory = listRecentForSource('diagnosis', { limit: 20 });

    return {
      ok: true,
      allHealthy: failing.length === 0,
      checkedJustNow: results,
      recentActivity: recentHistory.map((t) => ({ checkId: t.sourceRef, phase: t.phase, kind: t.kind, summary: t.summary, at: t.createdAt })),
      note:
        failing.length > 0
          ? 'One or more checks are failing right now — say so plainly, don\'t soften it, and name which ones.'
          : undefined,
    };
  },
};

File: backend/jarvis/tools/check_my_health.py
"""Has anything actually gone wrong with Jarvis itself lately?

Answered from real diagnostic runs and their traced outcomes — not from an
impression, and not from the absence of a memory of something going wrong. "I
think I'm fine" is exactly the answer this replaces.
"""

from __future__ import annotations

from typing import Any

from ..capabilities import CapabilitySpec, Risk


def _run(hours: int = 24) -> dict[str, Any]:
    from ..ops.diagnostics import checks, source

    checks.register_all()
    window = max(1, min(int(hours or 24), 24 * 14))
    health = source.recent_health(hours=window)

    failing = health["currentlyFailing"]
    if not health["recent"]:
        note = ("No checks have run yet, so there is nothing to report — say that, rather "
                "than saying everything is fine.")
    elif failing:
        note = "These are real, current failures, not a general impression."
    else:
        note = "Every check that has run recently passed."

    return {"ok": True, "hours": window, "currentlyFailing": failing,
            "checks": health["checksRegistered"], "recent": health["recent"][:20],
            "note": note}


SPEC = CapabilitySpec(
    id="builtin.check_my_health",
    name="check_my_health",
    description=("Check whether anything has actually gone wrong with Jarvis itself recently — "
                 "its database, memory, background work, scheduling, voice or security checks. "
                 "Use it whenever asked if you are working properly, or if something seems off "
                 "about your own behaviour. If nothing has run yet, say that rather than "
                 "claiming everything is fine."),
    input_schema={"type": "object", "properties": {
        "hours": {"type": "integer", "description": "How far back to look. Defaults to 24."}},
        "required": []},
    risk=Risk.LOW,
    handler=_run,
    timeout_s=15.0,
    tags=frozenset({"core", "meta"}),
)
